src/lib.rs

```diff
@@ -0,0 +1,489 @@
+    ///
+    /// The hot count for the desired location must be zero before calling this and you must be
+    /// very careful when using this function in multi-threaded tests. Two threads must not
+    /// concurrently call this function on the same location, otherwise the hot counts and JIT
+    /// transitions cannot be checked due to non-determinism.
+    fn dummy_trace_location(loc: JitLocation) {
+        // Interpreting phase.
+        for i in 0..(HOT_THRESHOLD - 1) {
+            assert!(hot_count(loc) == i);
+            assert!(JitThreadContext::current_phase() == JitPhase::Interpreting);
+            assert!(jit_control_point(loc) == JitTransition::IncrementedHotCount);
+        }
+
+        // The next iteration should trigger tracing.
+        assert!(hot_count(loc) == HOT_THRESHOLD - 1);
+        assert!(JitThreadContext::current_phase() == JitPhase::Interpreting);
+        assert!(jit_control_point(loc) == JitTransition::StartedTracing);
+
+        // When the start of the loop being traced shows up, we should fall back to the interpreter
+        // while the trace is being compiled.
+        assert!(JitThreadContext::current_phase() == JitPhase::Tracing(loc));
+        assert!(jit_control_point(loc) == JitTransition::FinishedTracing);
+        assert!(JitThreadContext::current_phase() == JitPhase::Interpreting);
+    }
+
+    // Continually call the JIT control point until a trace executes for the given location.
+    fn dummy_exec_trace(loc: JitLocation) {
+        let mut tries = 0;
+        let sleep_time = time::Duration::from_millis(20);
+        let max_tries = 1000;
+
+        loop {
+            match jit_control_point(loc) {
+                JitTransition::ExecutedTrace => break,  // The trace appeared and ran, good.
+                _ => {},                                // Still waiting.
+            }
+            tries += 1;
+            if tries == max_tries {
+                panic!("timed out waiting for trace to execute");
+            }
+            thread::sleep(sleep_time);
+        }
+    }
+
+    /// Test the most basic scenario.
+    #[test]
+    fn test_jit_0() {
+        reset_jit();
+        dummy_trace_location(666);
+        dummy_exec_trace(666);
+    }
+
+    /// Test tracing a few different locations.
+    #[test]
+    fn test_jit_lifecycle_1() {
+        reset_jit();
+        dummy_trace_location(666);
+        dummy_trace_location(777);
+        dummy_trace_location(888);
+        dummy_exec_trace(666);
+        dummy_exec_trace(777);
+        dummy_exec_trace(888);
+    }
+
+    /// Test tracing a few different locations in a weird order.
+    #[test]
+    fn test_jit_lifecycle_2() {
+        reset_jit();
+        dummy_trace_location(888);
+        dummy_trace_location(666);
+        dummy_exec_trace(888);
+        dummy_trace_location(777);
+        dummy_exec_trace(777);
+        dummy_exec_trace(666);
+    }
+
+    /// Guidelines for writing multi-threaded tests in this module.
```

Comment:
Surely these guidelines should be earlier? Also, I think they should be `//` not `///` (the latter is for "rustdoc" and doesn't really make sense in a test module).

Reply:
I put the threaded test comment before the first threaded test. Could haul it up if you like.